File: trunk/dotnet/nanocontainer/src/IntegrationKit/LifeCycleContainerBuilder.cs
using System;
using PicoContainer.Core;
using PicoContainer.Defaults;

namespace NanoContainer.IntegrationKit
{
	public abstract class LifeCycleContainerBuilder : ContainerBuilder
	{
		public void BuildContainer(IObjectReference containerRef,
		                           IObjectReference parentContainerRef,
		                           Object assemblyScope)
		{
			IMutablePicoContainer parent = parentContainerRef == null ? null : parentContainerRef.Get() as IMutablePicoContainer;
			IMutablePicoContainer container = CreateContainer(parent, assemblyScope);

			// register the child in the parent so that lifecycle can be propagated down the hierarchy
			if (parent != null)
			{
				parent.UnregisterComponentByInstance(container);
				parent.RegisterComponentInstance(container, container);
			}

			ComposeContainer(container, assemblyScope);
			container.Start();

			containerRef.Set(container);

		}

		public void KillContainer(IObjectReference containerRef)
		{
			IMutablePicoContainer pico = containerRef == null ? null : containerRef.Get() as IMutablePicoContainer;

			try
			{
				if (pico != null)
				{
					pico.Stop();
					pico.Dispose();

					IMutablePicoContainer parent = pico.Parent as IMutablePicoContainer;
					if (parent != null)
					{
						parent.UnregisterComponentByInstance(pico);
					}
				}
			}
			finally
			{
				pico = null;
			}
		}

		protected abstract void ComposeContainer(IMutablePicoContainer container, Object assemblyScope);

		protected abstract IMutablePicoContainer CreateContainer(IPicoContainer parentContainer,
		                                                         Object assemblyScope);
	}
}

File: trunk/java/nanocontainer/src/java/org/nanocontainer/integrationkit/LifecycleContainerBuilder.java
/*****************************************************************************
 * Copyright (C) NanoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 *****************************************************************************/
package org.picoextras.integrationkit;

import org.picocontainer.MutablePicoContainer;
import org.picocontainer.defaults.DefaultPicoContainer;
import org.picocontainer.defaults.ObjectReference;

/**
 * @author <a href="mailto:joe@thoughtworks.net">Joe Walnes</a>
 * @author Aslak Helles&oslash;y
 * @version $Revision$
 */
public abstract class LifecycleContainerBuilder implements ContainerBuilder {

    public void buildContainer(ObjectReference containerRef, ObjectReference parentContainerRef, Object assemblyScope) {

        MutablePicoContainer container = createContainer();

        if (parentContainerRef != null) {
            MutablePicoContainer parent = (MutablePicoContainer) parentContainerRef.get();
            container.setParent(parent);
        }

        composeContainer(container, assemblyScope);
        container.start();

        // hold on to it
        containerRef.set(container);
    }

    protected abstract void composeContainer(MutablePicoContainer container, Object assemblyScope);

    public void killContainer(ObjectReference containerRef){
        try {
            MutablePicoContainer pico = (MutablePicoContainer) containerRef.get();
            pico.stop();
            pico.dispose();
            pico.setParent(null);
        } finally {
            containerRef.set(null);
        }
    }

    protected MutablePicoContainer createContainer() {
		return new DefaultPicoContainer();
	}
}
